### ISSUE TITLE
`if (IS_DEBUG)` vs `#if IS_DEBUG`

### ISSUE THREAD

**Issue by mrzzzrm** [hyrise/hyrise#507]
Maintainers, please make a call. I think the second is more intuitive and more general, e.g. also applies to `#if NUMA_SUPPORTED`

Right now we 50/50 this in our code base.

**Comment by lanice** [comment]
I'd vote for `#if IS_DEBUG` as well.

**Comment by TwoLaid** [comment]
I think `#if`/`#ifdef` is great in headers for defining other macros and types, but in implementation code it always seems out of place for me. IDEs get confused about whether the macro is set or not and can screw up the highlighting.. 

Also, when you need an `#if` in a very nested code block, the if-condition is on the very left and far away from the execution block, like this:

```c++
namespace foo {
  // ...
  void bar() {
    // ...
    if (...) {
      // ...
#if IS_DEBUG
      do_debuggy_stuff();
#endif

      // vs.

      if (IS_DEBUG) {
        do_debuggy_stuff();
      }
    }
  }
}
```

Also, `if (IS_DEBUG)` nicely adds another level of indentation, which I think is correct because it adds another node in the AST.

TL;DR: `if (IS_DEBUG)`

**Comment by mrks** [comment]
I'm fine with either. Slight tendency towards regular ifs because that makes sure that even the removed code compiles. That is good because most of us don't develop on a NUMA system.

 Have you given `if constexpr` a thought? That doesn't have that advantage though.

**Comment by TwoLaid** [comment]
Another problem with `if constexpr` is that we need to use `// clang-format off/on` ..

**Comment by mrzzzrm** [comment]
"it always seems out of place for me" I think I'm a bit of a C++ conservative here, I have never seen `if (...)` but plenty of `#if ...`.

We already have `#if HYRISE_NUMA_SUPPORT` which encapsulates code that won't even link on other platforms. We should have one way of doing conditional compilation (duuuh, `if (...)` is not even real "conditional compilation") and not two. Indentation and improper IDE support (CLion does it just fine and even folds the unused code away) are not proper arguments imo, consistency within our code base and with other C++ projects on the other hand is.

**Comment by TwoLaid** [comment]
Just to clarify, before this project I've never seen `if ()` used with macros but I thought "wow that's really elegant and so much nicer than `#if`". `#if` always looked out of place for me (in implementation code) but I didn't know about the alternative

**Comment by lawben** [comment]
Even though I personally think the `if (IS_DEBUG)` looks nice, I think we should go with `#if IS_DEBUG` because it seems better to separate our actual conditions, which define major program flow, from some preprocessor steps that make sure slightly different code is produced. As our logic does not change with `IS_DEBUG` but only some additional checks are made, it should not be used like a 'regular' conditional.

**Comment by Bensk1** [comment]
I slightly prefer regular ifs, too and would go with all of @TwoLaid's arguments.

**Comment by mrks** [comment]
I'll change my opinion to a somewhat stronger feeling towards `#if`. Before I said that `if` would partially compile both paths before discarding the untaken one. This makes sure that everything can be build, but it can become a problem. For instance, it means that everyone has to have all NUMA libraries installed. If, heaven forbid, we need platform-specific code in the future, `if` would also mean all types of trouble.

**Comment by mrks** [comment]
Let's put an end to this. I talked to @mrzzzrm and @Bensk1 and we believe that the last argument mentioned for `#if` is the stronger one. I do agree that `if()` looks nicer, but `#if`s functionality is better.

**Comment by mrks** [comment]
Gnaaah. I just started to change everything to `#if` and then I found this:

```
NodeQueueScheduler::~NodeQueueScheduler() {
  if (IS_DEBUG && !_shut_down) {
    // We cannot throw an exception because destructors are noexcept by default.
    std::cerr << "NodeQueueScheduler::finish() wasn't called prior to destroying it" << std::endl;
    std::exit(EXIT_FAILURE);
  }
}
```

Changing that to `#if` makes the code less readable because we have another level of indentation.

I believe there are places where `#if` is better (platform-dependent code, NUMA support) and others where `if()` is better (IS_DEBUG). The compiled code is the same in any case. From a readability perspective, I expect everyone to be able to understand both versions without difficulties.

I'm all for harmonization and am a big fan of going over existing code and trying to improve it. In this case, I feel like we're wasting our time.
